stdlib/strtod.c

```diff
@@ -15,184 +15,539 @@
+				exp_max = LDBL_MAX_EXP + LDBL_MANT_DIG;
 			}
 			else {
-				result *= tmp;
+				exp_min = LDBL_MIN_10_EXP - LDBL_MANT_10_DIG;
+				exp_max = LDBL_MAX_10_EXP + LDBL_MANT_10_DIG;
 			}
+			break;
+	}
+
+
+	size_t overflow_int = 0, frac_position = 0;
+	long double result;
+	int64_t total_exp = 0; /* base 2 exponent if hex, base 10 exponent otherwise */
+
+	if (type <= TYPE_ID_DBL) {
+		/* If parsing singles/doubles it's sufficient to use uint64_t to store
+		   the digits - this allows us to use integer multiplication */
+		const uint64_t limit = hex ? (UINT64_MAX / 16) : (10000000000000000000ull / 10);
+		uint64_t digits;
+		str = scan_number(str, hex, decimal_point, limit, &frac_position, &overflow_int, &digits);
+		result = (long double)digits;
+	}
+	else {
+		str = scan_number_long_double(str, hex, decimal_point, LDBL_MANT_10_DIG, &frac_position, &overflow_int, &result);
+	}
 
-			is_number = 1;
+	if (frac_position > (INT32_MAX / 4) || overflow_int > (INT32_MAX / 4)) {
+		/* This is an absurd situation, we read in 2^29 digits from input, so we don't care much about the result */
+		total_exp = (frac_position != 0) ? INT32_MIN : INT32_MAX;
+	}
+	else {
+		/* If there was a decimal point, we don't need to care about the overflow */
+		total_exp = (frac_position != 0) ? -(int64_t)frac_position : overflow_int;
+		if (hex) {
+			total_exp *= 4;
+		}
+	}
+
+
+	bool has_exponent = hex ? (*str == 'P' || *str == 'p') : (*str == 'E' || *str == 'e');
+	if (has_exponent) {
+		const char *str_exp = str + 1;
+		bool exp_negative = false;
+		if ((*str_exp == '-') || (*str_exp == '+')) {
+			exp_negative = (*str_exp == '-');
+			str_exp++;
 		}
 
-		tmp *= tmp;
+		if (isdigit(*str_exp)) {
+			size_t exp_overflowed = 0, dummy;
+			uint64_t exp_scan;
+			str = scan_number(str_exp, 0, 0, ((INT32_MAX - 9) / 10), &dummy, &exp_overflowed, &exp_scan);
+			if (exp_overflowed != 0) {
+				total_exp = exp_negative ? INT32_MIN : INT32_MAX;
+			}
+			else {
+				/* At this point exp_scan is guaranteed to be <= INT32_MAX */
+				total_exp += exp_negative ? -(int32_t)exp_scan : exp_scan;
+			}
+		}
 	}
 
-	if (negative != 0) {
-		result = -result;
+	if (result != 0.0l) {
+		if (total_exp < exp_min) {
+			result = 0;
+		}
+		else if (total_exp > exp_max) {
+			result = INFINITY;
+		}
+		else if (hex) {
+			result = mul_exp2(result, total_exp);
+		}
+		else {
+			result = mul_exp10(result, total_exp);
+		}
+
+		if (result == 0.0l) {
+			/* We entered with non-zero result, but now it is zero - underflow must have occurred */
+			/* TODO: detect denormalization */
+			errno = ERANGE;
+		}
 	}
 
-	if (endptr != NULL) {
-		*endptr = (char *)str;
+	*out = result;
+	return str;
+}
+
+
+static long double strto_any(const char *__restrict str, char **__restrict endptr, uint8_t type)
+{
+	const char *decimal_point_str = localeconv()->decimal_point;
+	int decimal_point = decimal_point_str[0];
+	if ((decimal_point == 0) || (decimal_point_str[1] != 0)) {
```

Comment:
If the decimal point is not defined in the current locale, then `.` shall be used as a default (according to POSIX), instead of aborting the conversion.

Reply:
This check is not meant for when the decimal point is not defined, but rather for when it is invalid. For example, glibc checks for this case using an assertion: https://github.com/lattera/glibc/blob/895ef79e04a953cac1493863bcae29ad85657ee1/stdlib/strtod_l.c#L592